<?xml version="1.0"?>
<launch>

  <!--Arguments-->
  <arg name="r_name" default="racecar"/>
  <arg name="vesc_ns" default="/vesc"/>

  <arg name="r_type1" default="racecar"/>
  <arg name="max_vel1" default="0.5"/>
  <arg name="max_acc1" default="0.1"/>

  <arg name="r_type2" default="racecar"/>
  <arg name="max_vel2" default="0.5"/>
  <arg name="max_acc2" default="0.1"/>

  <arg name="r_type3" default="racecar"/>
  <arg name="max_vel3" default="0.5"/>
  <arg name="max_acc3" default="0.1"/>

  <arg name="r_type4" default="racecar"/>
  <arg name="max_vel4" default="0.5"/>
  <arg name="max_acc4" default="0.1"/>

  <arg name="comm_range" default="10"/>

  <param name="robot1_num" value="1"/>
  <param name="r_type1" value="$(arg r_type1)"/>
  <param name="max_vel1" value="$(arg max_vel1)"/>
  <param name="max_acc1" value="$(arg max_acc1)"/>

  <param name="robot2_num" value="2"/>
  <param name="r_type2" value="$(arg r_type2)"/>
  <param name="max_vel2" value="$(arg max_vel2)"/>
  <param name="max_acc2" value="$(arg max_acc2)"/>

  <param name="robot3_num" value="3"/>
  <param name="r_type3" value="$(arg r_type3)"/>
  <param name="max_vel3" value="$(arg max_vel3)"/>
  <param name="max_acc3" value="$(arg max_acc3)"/>

  <param name="robot4_num" value="4"/>
  <param name="r_type4" value="$(arg r_type4)"/>
  <param name="max_vel4" value="$(arg max_vel4)"/>
  <param name="max_acc4" value="$(arg max_acc4)"/>

  <param name="comm_range" value="$(arg comm_range)"/>

  <!--Each Racecar Broadcasts its Identification Message-->

  <node pkg="racecar_communication" type="send_id_msg.py" name="id_publisher1" output="screen">
    <remap from="/vesc/odom" to="$(arg vesc_ns)1/odom" />
    <remap from="/move_base/local_costmap/footprint" to="$(arg r_name)1/move_base/local_costmap/footprint" />
    <remap from="robot_num" to="robot1_num"/>
    <remap from="r_type" to="r_type1"/>
    <remap from="max_vel" to="max_vel1"/>
    <remap from="max_acc" to="max_acc1"/>
  </node>

  <node pkg="racecar_communication" type="send_id_msg.py" name="id_publisher2" output="screen">
    <remap from="/vesc/odom" to="$(arg vesc_ns)2/odom" />
    <remap from="/move_base/local_costmap/footprint" to="$(arg r_name)2/move_base/local_costmap/footprint" />
    <remap from="robot_num" to="robot2_num"/>
    <remap from="r_type" to="r_type2"/>
    <remap from="max_vel" to="max_vel2"/>
    <remap from="max_acc" to="max_acc2"/>
  </node>

  <node pkg="racecar_communication" type="send_id_msg.py" name="id_publisher3" output="screen">
    <remap from="/vesc/odom" to="$(arg vesc_ns)3/odom" />
    <remap from="/move_base/local_costmap/footprint" to="$(arg r_name)3/move_base/local_costmap/footprint" />
    <remap from="robot_num" to="robot3_num"/>
    <remap from="r_type" to="r_type3"/>
    <remap from="max_vel" to="max_vel3"/>
    <remap from="max_acc" to="max_acc3"/>
  </node>

  <node pkg="racecar_communication" type="send_id_msg.py" name="id_publisher4" output="screen">
    <remap from="/vesc/odom" to="$(arg vesc_ns)4/odom" />
    <remap from="/move_base/local_costmap/footprint" to="$(arg r_name)4/move_base/local_costmap/footprint" />
    <remap from="robot_num" to="robot4_num"/>
    <remap from="r_type" to="r_type4"/>
    <remap from="max_vel" to="max_vel4"/>
    <remap from="max_acc" to="max_acc4"/>
  </node>

  <!--Each Racecar Receives Others' Identification Messages-->
  <node pkg="racecar_communication" type="rec_id_msg.py" name="id_subscriber1" output="screen">
    <remap from="robot_num" to="robot1_num"/>
    <remap from="/vesc/odom" to="$(arg vesc_ns)1/odom" />
    <remap from="/ids_combined" to="/$(arg r_name)1/ids_combined" />
    <remap from="/footprints_combined" to="/$(arg r_name)1/footprints_combined" />
  </node>

  <node pkg="racecar_communication" type="rec_id_msg.py" name="id_subscriber2" output="screen">
    <remap from="robot_num" to="robot2_num"/>
    <remap from="/vesc/odom" to="$(arg vesc_ns)2/odom" />
    <remap from="/ids_combined" to="/$(arg r_name)2/ids_combined" />
    <remap from="/footprints_combined" to="/$(arg r_name)2/footprints_combined" />
  </node>

  <node pkg="racecar_communication" type="rec_id_msg.py" name="id_subscriber3" output="screen">
    <remap from="robot_num" to="robot3_num"/>
    <remap from="/vesc/odom" to="$(arg vesc_ns)3/odom" />
    <remap from="/ids_combined" to="/$(arg r_name)3/ids_combined" />
    <remap from="/footprints_combined" to="/$(arg r_name)3/footprints_combined" />
  </node>

  <node pkg="racecar_communication" type="rec_id_msg.py" name="id_subscriber4" output="screen">
    <remap from="robot_num" to="robot4_num"/>
    <remap from="/vesc/odom" to="$(arg vesc_ns)4/odom" />
    <remap from="/ids_combined" to="/$(arg r_name)4/ids_combined" />
    <remap from="/footprints_combined" to="/$(arg r_name)4/footprints_combined" />
  </node>

</launch>
